### ISSUE TITLE
PyLance reports possibly missing key when its existence can be statically inferred

### ISSUE THREAD

**Issue by fcole90** [microsoft/pyright#3683]
## Environment data

-   Language Server version: 2022.7.20 (pyright 7cadfb74)
-   OS and version: Ubuntu 22.04 LTS
-   Python version: 3.10.4

## Code Snippet

`src/test.py`:
```python
from typing import TypedDict


class PartialCustomDict(TypedDict, total=False):
    myProp: str


def main():
    pdc: PartialCustomDict = {}
    pdc["myProp"] = "test"
    _ = pdc["myProp"]


if __name__ == "__main__":
    main()
```

## Repro Steps

`.vscode/settings.json`:
```json
{
  "python.analysis.typeCheckingMode": "strict",
  "python.analysis.logLevel": "Trace",
  "python.linting.flake8Enabled": true,
  "python.linting.enabled": true,
  "python.linting.mypyEnabled": false,
  "python.linting.pylintEnabled": false
}
```

## Expected behavior
No error should be reported attempting to access `pdc["myProp"]`

## Actual behavior
PyLance complains that `pyProp` might not exist, even if it's clear that it must exist at that point.

```
Could not access item in TypedDict
  "myProp" is not a required key in "PartialCustomDict", so access may result in runtime exceptionPylance[reportTypedDictNotRequiredAccess](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)
```

## Logs

*There was an error creating your Issue: body is too long (maximum is 65536 characters).*

**Comment by erictraut** [comment]
The current behavior is "as designed", not a bug. But I can understand why this error would be unexpected.

Pyright contains some special-case logic that narrows the type of a TypedDict on assignment from a literal dictionary expression. If a key is "not required", it is marked as "known present". That's why your code sample works if you initialize the variable with the prop. 

To make the error go away, pyright would need to perform further type narrowing on assignments of keys (and for completeness, deletions of keys as well). Currently that type narrowing logic doesn't exist.

I think this is a reasonable feature request, but it will take a fair amount of work to get it right.

**Comment by fcole90** [comment]
I get it, thank you for your detailed explanation :blush: Yes, with a background in typescript, this came very surprising, but I understand that PyLance is a much more recent project than TS :blush:

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.259, which I just published. It will also be included in the next release of pylance.

**Comment by Seairth** [comment]
> This is addressed in pyright 1.1.259, which I just published. It will also be included in the next release of pylance.

Can you provide a link to this information?  Or, if this is now a configurable rule, a link to the configuration information?

**Comment by debonte** [comment]
> Can you provide a link to this information? Or, if this is now a configurable rule, a link to the configuration information?

@Seairth, I don't believe there's any configuration required here. Using the OP's code above and strict mode, I don't see any diagnostics in the latest Pyright build or the latest Pylance build. Given that this issue is 18 months old, if you're seeing an unexpected diagnostic related to TypedDicts, you should file a new issue.

**Comment by Seairth** [comment]
> > Can you provide a link to this information? Or, if this is now a configurable rule, a link to the configuration information?
> 
> @Seairth, I don't believe there's any configuration required here. Using the OP's code above and strict mode, I don't see any diagnostics in the latest Pyright build or the latest Pylance build. Given that this issue is 18 months old, if you're seeing an unexpected diagnostic related to TypedDicts, you should file a new issue.

Thanks.  mine is a slightly different situation, where I'm returning the dict from a cpython library, where I know that all the `NotRequired` fields set/present.  In this case, the error message isn't wrong, because the code in question can't know that.  Still, I was hoping there was a flag to simply turn off this rule.  I know it's not the ideal solution, but the other solutions I can think of would significantly clutter the code solely to satisfy pylance/pyright.

**Comment by debonte** [comment]
@Seairth, I see. You can disable individual diagnostic rules globally by setting their severity to `false` or `"none"` in your config. The list of rules is [here](https://microsoft.github.io/pyright/#/configuration?id=type-check-diagnostics-settings) if you need it. And there's a sample `pyrightconfig.json` file [here](https://microsoft.github.io/pyright/#/configuration?id=sample-config-file) that sets the severity for a couple of rules. It's around the middle of the file -- look for `"reportMissingTypeStubs"`.

You can also change the severity of diagnostic rules for a particular file or suppress particular diagnostics on a single line as described [here](https://microsoft.github.io/pyright/#/comments).
